source/D2Lang/src/D2Unicode/D2UnicodeChar.cpp

```diff
@@ -36,3 +36,104 @@ Unicode& Unicode::operator=(const Unicode& str) {
 Unicode::operator unsigned short() const {
   return this->ch;
 }
+
+Unicode::Direction Unicode::directionality() {
+  /*
+   * Hebrew, Arabic, Syriac, Arabic Supplement, Thaana, NKo,
+   * Samaritan, Mandaic, Syriac Supplement, Arabic Extended-A
+   */
+  if (this->ch >= 0x590 && this->ch <= 0x8FF) {
+    return DIR_RIGHT_TO_LEFT;
+  }
+
+  /* Lower case a to z */
+  if (this->ch >= L'a' && this->ch <= L'z') {
+    return DIR_LEFT_TO_RIGHT;
+  }
+
+  /* @ character, upper case A to Z */
+  if (this->ch >= L'@' && this->ch <= L'Z') {
+    return DIR_LEFT_TO_RIGHT;
+  }
+
+  /* Numbers 0 to 9 */
+  if (this->ch >= L'0' && this->ch <= L'9') {
+    return DIR_LEFT_TO_RIGHT;
+  }
+
+  /*
+   * The characters \0 to [SPACE] and the following:
+   * !"#$%'()*+,-./:;<=>?
+   *
+   * & character is a left-to-right case
+   */
+  if (this->ch <= L'\?') {
```

Comment:
Is there a need for the escape here ? why `\?` and not just `?` ?

Reply:
The question mark was added out of habit, because of the way that [trigraphs can affect how string literals containing `"??"` are interpretted](https://en.cppreference.com/w/c/language/escape). In this case, the use of `'\?'` instead of `'?'` would be unnecessary, since only one character is intended to be used in this comparison.